File: js/Output.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var THREE = require("three");
var Common_1 = require("./Common");
var Simulation_1 = require("./Simulation");
var face_vert_raw_1 = require("../shaders/face.vert?raw");
var color_frag_raw_1 = require("../shaders/color.frag?raw");
var Output = /** @class */ (function () {
    function Output() {
        this.simulation = new Simulation_1.default();
        this.scene = new THREE.Scene();
        this.camera = new THREE.Camera();
        this.output = new THREE.Mesh(new THREE.PlaneBufferGeometry(2, 2), new THREE.RawShaderMaterial({
            vertexShader: face_vert_raw_1.default,
            fragmentShader: color_frag_raw_1.default,
            uniforms: {
                velocity: {
                    value: this.simulation.fbos.vel_0.texture,
                },
                boundarySpace: {
                    value: new THREE.Vector2(),
                },
            },
        }));
        this.scene.add(this.output);
    }
    Output.prototype.addScene = function (mesh) {
        this.scene.add(mesh);
    };
    Output.prototype.resize = function () {
        this.simulation.resize();
    };
    Output.prototype.render = function () {
        Common_1.default.renderer.setRenderTarget(null);
        Common_1.default.renderer.render(this.scene, this.camera);
    };
    Output.prototype.update = function () {
        this.simulation.update();
        this.render();
    };
    return Output;
}());
exports.default = Output;

File: src/Output.lua
local Common = require("src.Common")
local THREE = js.global.THREE

local Simulation = require("src.Simulation")
local Object = require("utils.convertToJSObject")
local new = require("utils.new")
local get = require("utils.shaders")
local face_vert = get("face.vert")
local color_frag = get("color.frag")

return function()
    local self = {}

    self.simulation = new(Simulation)

    self.scene = new THREE.Scene()
    self.camera = new THREE.Camera()

    self.output = new(THREE.Mesh,
        new(THREE.PlaneBufferGeometry2, 2),
        new(THREE.RawShaderMaterial, Object({
            vertexShader = face_vert,
            fragmentShader = color_frag,
            uniforms = {
                velocity = {
                    value = self.simulation.fbos.vel_0.texture
                },
                boundarySpace = {
                    value = new(THREE.Vector2)
                }
            }
        }))
    )

    self.scene.add(self.output)

    self.addScene = function(mesh)
        self.scene.add(mesh)
    end

    self.resize = function()
        self.simulation.resize()
    end

    self.render = function()
        Common.renderer.setRenderTarget(nil)
        Common.renderer.render(self.scene, self.camera)
    end

    self.update = function()
        self.simulation.update()
        self.render()
    end

    print("Output.lua initialized")
    return self
end